File: pkg/earthsci-astdiff-py/src/earthsci_astdiff/numeric.py
"""Numeric assembly of the band entries — a correctness surface, not a
performance one.

The Julia reference compiles a derived evaluation model through its
tree-walk; here each band coefficient is evaluated per (row, contracted)
cell by resolving every read (state/parameter/`t`/const-table gathers and
scalar ``reduce: "+"`` aggregates) to a literal and handing the closed tree
to the official ``numpy_interpreter`` evaluator — so operator semantics
(ifelse, min/max, ``fn`` lookups) stay the runtime's own. Complexity is
O(cells × coefficient size) per call; use the Julia binding for production
assembly.
"""
from __future__ import annotations

from itertools import product

from earthsci_ast.esm_types import Expr, ExprNode
from earthsci_ast.expr_walk import map_children
from earthsci_ast.numpy_interpreter import evaluate
from earthsci_ast.substitute import substitute

from .emit import _cellname, _eval_cidx, _product
from .jacobian import jacobian_bands
from .system import SysView, _var_type, ctx_of, sysview


class EvalError(Exception):
    pass


class _Env:
    """Point bindings: state arrays (nested lists, 1-based wire indices),
    scalar/array parameters, and the time value."""

    def __init__(self, states: dict, params: dict, t: float):
        self.states = states
        self.params = params
        self.t = t

    def scalar(self, name: str):
        if name == "t":
            return self.t
        if name in self.params and not isinstance(self.params[name], list):
            return self.params[name]
        if name in self.states and not isinstance(self.states[name], list):
            return self.states[name]
        return None

    def gather(self, name: str, cells: list[int]):
        arr = self.states.get(name, self.params.get(name))
        if arr is None:
            return None
        for c in cells:
            arr = arr[c - 1]
        return arr


def _resolve(e: Expr, env: _Env) -> Expr:
    """Fold every read in ``e`` to a literal (recursively), leaving a closed
    tree the numpy evaluator can run with no bindings."""
    if isinstance(e, str):
        v = env.scalar(e)
        if v is None:
            raise EvalError(f"unbound name `{e}` in coefficient")
        return v
    if not isinstance(e, ExprNode):
        return e
    if e.op == "index":
        t = e.args[0]
        cells = [int(round(float(_as_lit(_resolve(x, env)))))
                 for x in e.args[1:]]
        if isinstance(t, str):
            v = env.gather(t, cells)
            if v is None:
                raise EvalError(f"index into unbound array `{t}`")
            return v
        if isinstance(t, ExprNode) and t.op == "const":
            v = t.value
            for c in cells:
                v = v[c - 1]
            return v
        raise EvalError(f"index into unsupported target `{t}`")
    if e.op == "aggregate":
        if e.output_idx:
            raise EvalError("array-producing aggregate in scalar coefficient")
        if e.reduce not in (None, "+"):
            raise EvalError(f"reduce=`{e.reduce}` in scalar coefficient")
        names = sorted(str(n) for n in (e.ranges or {}))
        boxes = [range(int(e.ranges[n][0]), int(e.ranges[n][1]) + 1)
                 for n in names]
        total = 0.0
        for cell in product(*boxes):
            body = substitute(e.expr, {n: cell[k]
                                       for k, n in enumerate(names)})
            total += float(_as_lit(_resolve(body, env)))
        return total
    return map_children(e, lambda x: _resolve(x, env))


def _as_lit(e: Expr) -> float:
    if isinstance(e, (int, float)) and not isinstance(e, bool):
        return float(e)
    return float(evaluate(e, {}))


def _default_values(sv: SysView, ctx):
    states: dict = {}
    params: dict = {}
    for n, var in sv.variables.items():
        ty = _var_type(var)
        if ty not in ("state", "parameter"):
            continue
        default = getattr(var, "default", None)
        shape = ctx.shapes.get(str(n))
        if shape:
            val = _nested_full(shape, 0.0 if default is None else float(default))
        else:
            val = 0.0 if default is None else float(default)
        (states if ty == "state" else params)[str(n)] = val
    return states, params


def _nested_full(shape: list[int], v: float):
    if len(shape) == 1:
        return [v] * shape[0]
    return [_nested_full(shape[1:], v) for _ in range(shape[0])]


def assemble_jacobian(obj, model_name: str | None = None, wrt: str = "states",
                      states: dict | None = None, params: dict | None = None,
                      t: float = 0.0) -> dict[tuple[str, str], float]:
    """Evaluate the band entries at a point: ``{(rowname, colname): value}``
    with contributions to one (row, col) pair SUMMED (§4 additivity, incl.
    contracted-point accumulation). ``states``/``params`` map variable name
    → nested list (1-based cells) or scalar; defaults come from the document."""
    sv = obj if isinstance(obj, SysView) else sysview(obj, model_name)
    ctx = ctx_of(sv)
    dstates, dparams = _default_values(sv, ctx)
    if states:
        dstates.update(states)
    if params:
        dparams.update(params)
    env = _Env(dstates, dparams, t)
    entries = jacobian_bands(sv, wrt=wrt)
    out: dict[tuple[str, str], float] = {}
    for en in entries:
        b = en.band
        for cell in _product(b.rows):
            for cc in _product([(lo, hi) for _, lo, hi in b.contracted]):
                bind = {r: cell[k] for k, r in enumerate(b.ridx)
                        if isinstance(r, str)}
                for k, (nm, _, _) in enumerate(b.contracted):
                    bind[nm] = cc[k]
                fbind = {k2: float(v2) for k2, v2 in bind.items()}
                cidx = [_eval_cidx(c, fbind) for c in b.cidx]
                vshape = ctx.shapes.get(en.v)
                if vshape is not None and any(
                        not (1 <= cidx[d] <= vshape[d])
                        for d in range(len(cidx))):
                    continue                # ghost/boundary read: no column
                coef = substitute(b.coef, dict(bind)) if bind else b.coef
                val = float(_as_lit(_resolve(coef, env)))
                key = (_cellname(en.u, cell), _cellname(en.v, cidx))
                out[key] = out.get(key, 0.0) + val
    return out

File: pkg/earthsci-astdiff-rs/src/numeric.rs
//! Numeric assembly — a correctness surface (like the Python binding's):
//! per-cell read resolution to literals, then the official evaluator.

use std::collections::{BTreeMap, HashMap};

use earthsci_ast::expression::evaluate;
use earthsci_ast::substitute::substitute;
use earthsci_ast::types::{EsmFile, Expr};
use serde_json::Value;

use crate::bands::BandError;
use crate::clip::RIdx;
use crate::helpers::map_expr_children;
use crate::jacobian::{jacobian_bands, Wrt};
use crate::system::{ctx_of, sysview};

type R<T> = Result<T, BandError>;

pub struct Env {
    /// name → nested array (1-based cells encoded as serde Value) or scalar
    pub states: HashMap<String, Value>,
    pub params: HashMap<String, Value>,
    pub t: f64,
}

fn gather<'a>(v: &'a Value, cells: &[i64]) -> Option<&'a Value> {
    let mut cur = v;
    for c in cells {
        cur = cur.as_array()?.get((*c - 1) as usize)?;
    }
    Some(cur)
}

fn as_lit(e: &Expr) -> R<f64> {
    match e {
        Expr::Integer(i) => Ok(*i as f64),
        Expr::Number(x) => Ok(*x),
        _ => evaluate(e, &HashMap::new())
            .map_err(|er| BandError(format!("evaluate: {er:?}"))),
    }
}

fn resolve(e: &Expr, env: &Env) -> R<Expr> {
    match e {
        Expr::Variable(name) => {
            if name == "t" {
                return Ok(Expr::Number(env.t));
            }
            for m in [&env.params, &env.states] {
                if let Some(v) = m.get(name) {
                    if let Some(x) = v.as_f64() {
                        return Ok(Expr::Number(x));
                    }
                }
            }
            Err(BandError(format!("unbound name `{name}` in coefficient")))
        }
        Expr::Operator(n) if n.op == "index" => {
            let mut cells = Vec::new();
            for x in &n.args[1..] {
                cells.push(as_lit(&resolve(x, env)?)?.round() as i64);
            }
            match n.args.first() {
                Some(Expr::Variable(t)) => {
                    let arr = env.states.get(t).or_else(|| env.params.get(t))
                        .ok_or_else(|| BandError(format!(
                            "index into unbound array `{t}`")))?;
                    let v = gather(arr, &cells).ok_or_else(|| BandError(
                        format!("out-of-range gather into `{t}`")))?;
                    Ok(Expr::Number(v.as_f64().ok_or_else(|| BandError(
                        "non-numeric gather".into()))?))
                }
                Some(Expr::Operator(cn)) if cn.op == "const" => {
                    let v = cn.value.as_ref().ok_or_else(|| BandError(
                        "const without value".into()))?;
                    let v = gather(v, &cells).ok_or_else(|| BandError(
                        "out-of-range const gather".into()))?;
                    Ok(Expr::Number(v.as_f64().ok_or_else(|| BandError(
                        "non-numeric const gather".into()))?))
                }
                _ => Err(BandError("index into unsupported target".into())),
            }
        }
        Expr::Operator(n) if n.op == "aggregate" => {
            if n.output_idx.as_ref().map(|o| !o.is_empty()).unwrap_or(false) {
                return Err(BandError(
                    "array-producing aggregate in scalar coefficient".into()));
            }
            if let Some(red) = &n.reduce {
                if red != "+" {
                    return Err(BandError(format!(
                        "reduce=`{red}` in scalar coefficient")));
                }
            }
            let empty = HashMap::new();
            let ranges = n.ranges.as_ref().unwrap_or(&empty);
            let mut names: Vec<String> = ranges.keys().cloned().collect();
            names.sort();
            let ctx = crate::bands::Ctx::default();
            let boxes: Vec<(i64, i64)> = names.iter()
                .map(|nm| crate::bands::range_of(&ranges[nm], &ctx))
                .collect::<Result<_, _>>()?;
            let body = n.expr.as_deref().ok_or_else(|| BandError(
                "aggregate without body".into()))?;
            let mut total = 0.0;
            let mut cell: Vec<i64> = boxes.iter().map(|(lo, _)| *lo).collect();
            'outer: loop {
                let binds: HashMap<String, Expr> = names.iter().cloned()
                    .zip(cell.iter().map(|c| Expr::Integer(*c))).collect();
                total += as_lit(&resolve(&substitute(body, &binds), env)?)?;
                let mut k = 0;
                loop {
                    if k == cell.len() {
                        break 'outer;
                    }
                    cell[k] += 1;
                    if cell[k] <= boxes[k].1 {
                        break;
                    }
                    cell[k] = boxes[k].0;
                    k += 1;
                }
                if cell.is_empty() {
                    break;
                }
            }
            Ok(Expr::Number(total))
        }
        Expr::Operator(_) => {
            let mut err: Option<BandError> = None;
            let out = map_expr_children(e, &mut |x| match resolve(x, env) {
                Ok(v) => v,
                Err(er) => {
                    if err.is_none() {
                        err = Some(er);
                    }
                    x.clone()
                }
            });
            match err {
                Some(er) => Err(er),
                None => Ok(out),
            }
        }
        _ => Ok(e.clone()),
    }
}

fn cellname(v: &str, cell: &[i64]) -> String {
    if cell.is_empty() {
        v.to_string()
    } else {
        format!("{v}[{}]", cell.iter().map(|c| c.to_string())
                .collect::<Vec<_>>().join(","))
    }
}

fn fold_const_gathers(e: &Expr, bind: &HashMap<String, f64>) -> R<Expr> {
    let Expr::Operator(n) = e else { return Ok(e.clone()) };
    if n.op == "index" {
        if let Some(Expr::Operator(cn)) = n.args.first() {
            if cn.op == "const" {
                let mut v = cn.value.as_ref().ok_or_else(|| BandError(
                    "const without value".into()))?;
                for x in &n.args[1..] {
                    let i = eval_cidx(x, bind)?;
                    v = v.as_array().and_then(|a| a.get((i - 1) as usize))
                        .ok_or_else(|| BandError(
                            "out-of-range const gather".into()))?;
                }
                return Ok(match v {
                    Value::Number(num) if num.is_i64() =>
                        Expr::Integer(num.as_i64().unwrap()),
                    _ => Expr::Number(v.as_f64().ok_or_else(|| BandError(
                        "non-numeric const gather".into()))?),
                });
            }
        }
    }
    let mut err: Option<BandError> = None;
    let out = map_expr_children(e, &mut |x| {
        match fold_const_gathers(x, bind) {
            Ok(v) => v,
            Err(er) => {
                if err.is_none() {
                    err = Some(er);
                }
                x.clone()
            }
        }
    });
    match err {
        Some(er) => Err(er),
        None => Ok(out),
    }
}

fn eval_cidx(c: &Expr, bind: &HashMap<String, f64>) -> R<i64> {
    let folded = fold_const_gathers(c, bind)?;
    evaluate(&folded, bind)
        .map(|v| v.round() as i64)
        .map_err(|er| BandError(format!("evaluate cidx: {er:?}")))
}

fn product_first_fastest(ranges: &[(i64, i64)]) -> Vec<Vec<i64>> {
    let mut out = Vec::new();
    let mut cell: Vec<i64> = ranges.iter().map(|(lo, _)| *lo).collect();
    if ranges.is_empty() {
        return vec![vec![]];
    }
    loop {
        out.push(cell.clone());
        let mut k = 0;
        loop {
            if k == cell.len() {
                return out;
            }
            cell[k] += 1;
            if cell[k] <= ranges[k].1 {
                break;
            }
            cell[k] = ranges[k].0;
            k += 1;
        }
    }
}

/// Evaluate the band entries at a point: `{(rowname, colname): value}`,
/// contributions summed (§4 additivity, contracted points included).
pub fn assemble_jacobian(file: &EsmFile, model_name: Option<&str>, wrt: Wrt,
                         env: &Env) -> R<BTreeMap<(String, String), f64>> {
    let entries = jacobian_bands(file, model_name, wrt)?;
    let sv = sysview(file, model_name)?;
    let ctx = ctx_of(&sv)?;
    let mut out: BTreeMap<(String, String), f64> = BTreeMap::new();
    for en in &entries {
        let b = &en.band;
        for cell in product_first_fastest(&b.rows) {
            let cranges: Vec<(i64, i64)> =
                b.contracted.iter().map(|(_, lo, hi)| (*lo, *hi)).collect();
            for cc in product_first_fastest(&cranges) {
                let mut ibind: HashMap<String, i64> = HashMap::new();
                for (k, r) in b.ridx.iter().enumerate() {
                    if let RIdx::Name(nm) = r {
                        ibind.insert(nm.clone(), cell[k]);
                    }
                }
                for (k, (nm, _, _)) in b.contracted.iter().enumerate() {
                    ibind.insert(nm.clone(), cc[k]);
                }
                let fbind: HashMap<String, f64> =
                    ibind.iter().map(|(k, v)| (k.clone(), *v as f64)).collect();
                let cidx: Vec<i64> = b.cidx.iter()
                    .map(|c| eval_cidx(c, &fbind))
                    .collect::<Result<_, _>>()?;
                if let Some(vshape) = ctx.shapes.get(&en.v) {
                    if cidx.iter().zip(vshape.iter())
                        .any(|(c, s)| *c < 1 || c > s) {
                        continue;              // ghost read: no column
                    }
                }
                let coef = if ibind.is_empty() {
                    b.coef.clone()
                } else {
                    let sub: HashMap<String, Expr> = ibind.iter()
                        .map(|(k, v)| (k.clone(), Expr::Integer(*v))).collect();
                    substitute(&b.coef, &sub)
                };
                let val = as_lit(&resolve(&coef, env)?)?;
                let key = (cellname(&en.u, &cell), cellname(&en.v, &cidx));
                *out.entry(key).or_insert(0.0) += val;
            }
        }
    }
    Ok(out)
}
